url47af6b1535_rolandotr_db_comparison_tgz-pJ8-methodology_HistoryJ8/src/protocols/specifications/PoulidorProtocol.java

```diff
@@ -0,0 +1,290 @@
+			p.setNumberOfRounds(i);
+			BigDecimal mafia = p.getMafiaFraudProbability();
+			BigDecimal distance = p.getDistanceFraudProbability();
+			writerMafia.write(i+" "+mafia.doubleValue()+newLine);
+			writerDistance.write(i+" "+distance.doubleValue()+newLine);
+			System.out.println("n = "+i+" Mafia = "+mafia.doubleValue());
+			System.out.println("n = "+i+" Distance = "+distance.doubleValue());
+			writerMafia.flush();
+			writerDistance.flush();
+		}
+		writerMafia.close();
+		writerDistance.close();
+	}
+	
+	
+	public String getMafiaFileNameOfValues(){
+		return "mafia_"+getAcronym()+"_values.DAT";
+	}
+	public String getDistanceFileNameOfValues(){
+		return "distance_"+getAcronym()+"_values.DAT";
+	}
+
+	public static void main(String[] args) {
+		PoulidorProtocol p = new PoulidorProtocol();
+		p.setNumberOfRounds(256);
+		BigDecimal mafia = p.getMafiaFraudProbability();
+		System.out.println("n = "+256+" Mafia = "+mafia.doubleValue());
+		BigDecimal distance = p.getDistanceFraudProbability();
+		System.out.println("n = "+256+" Distance = "+distance.doubleValue());
+		//generateDataFile();
+	}
+
+	@Override
+	public int getTotalMsgSizeReceived() {
+		return sizeOfNonces+n;
+	}
+
+	@Override
+	public int getTotalOutputSizeOfCallsToFunctions() {
+		return 4*n;
+	}
+
+	@Override
+	public int getNumberOfNoncesGenerated() {
+		return 1;
+	}
+
+	@Override
+	public int getBitsGenerated() {
+		return 0;
+	}
+
+	private double[] mafia2;
+
+	private double mafiaFraudFromDisk2(int n){
+		if (mafia2 == null) {
+			try {
+				fillMafia2(257);
+			} catch (IOException e) {
+				// TODO Auto-generated catch block
+				e.printStackTrace();
+				throw new RuntimeException(e);
+			}
+		}
+		return mafia2[n];
+	}
+
+	/*La informacion sobre el distance esta en un fichero de nombre "d_FDA-"+n*/
+	private void fillMafia2(int n) throws IOException {
+		String tmp;
+		String[] split;
+		int r;
+		mafia2 = new double[n+1];
+		for (int i = 1; i <= n; i++){
+			File f = new File(this.getMafiaFileNameOfValues());
+			BufferedReader reader = new BufferedReader(new FileReader(f));
+			while (reader.ready()){
+				tmp = reader.readLine();
+				split = tmp.split(" ");
+				mafia2[Integer.parseInt(split[0])] = Double.parseDouble(split[1]);
+			}
+			reader.close();
+		}
+	}	
+
+	private double[] distances2;
+
+	private double distanceFraudFromDisk2(int n){
+		if (distances2 == null) {
+			try {
+				fillDistance2(257);
+			} catch (IOException e) {
+				// TODO Auto-generated catch block
+				e.printStackTrace();
+				throw new RuntimeException(e);
+			}
+		}
+		return distances2[n];
+	}
+
+	/*La informacion sobre el distance esta en un fichero de nombre "d_FDA-"+n*/
+	private void fillDistance2(int n) throws IOException {
+		String tmp;
+		String[] split;
+		int r;
+		distances2 = new double[n+1];
+		for (int i = 1; i <= n; i++){
+			File f = new File(this.getDistanceFileNameOfValues());
+			BufferedReader reader = new BufferedReader(new FileReader(f));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *reader*. The resource is referenced at lines 279, 280, 281. The resource is closed at line 285. There are other execution paths that do not contain closure statements, for example, when BufferedReader.ready throws an exception. Ensure *reader* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).